modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php

```diff
@@ -0,0 +1,247 @@
+<?php
+/**
+ * Background process runner class.
+ *
+ * @since n.e.x.t
+ * @package performance-lab
+ */
+
+/**
+ * Class Perflab_Background_Process.
+ *
+ * Runs the heavy lifting tasks in background in a separate process.
+ *
+ * @since n.e.x.t
+ */
+class Perflab_Background_Process {
+	/**
+	 * Name of the ajax action which will trigger background process to run the job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @var string
+	 */
+	const BG_PROCESS_ACTION = 'perflab_background_process_handle_request';
+
+	/**
+	 * Job instance.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @var Perflab_Background_Job|null
+	 */
+	private $job;
+
+	/**
+	 * Perflab_Background_Process constructor.
+	 *
+	 * Adds the necessary hooks to trigger process handling.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function __construct() {
+		// Handle job execution request.
+		add_action( 'wp_ajax_' . Perflab_Background_Process::BG_PROCESS_ACTION, array( $this, 'handle_request' ) );
+		add_action( 'wp_ajax_nopriv_' . Perflab_Background_Process::BG_PROCESS_ACTION, array( $this, 'handle_request' ) );
+	}
+
+	/**
+	 * Handles incoming request to run the batch of job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @throws Exception Invalid request for background process.
+	 */
+	public function handle_request() {
```

Comment:
@ankitrox I think this method does too much and should only be responsible for handling the HTTP request and calling the `run()` method. I would make the following changes:

- Remove the try/catch block and exceptions - as we're handling the HTTP request we can return early with a error response. This could be a combination of using `WP_Error` instead of `Exception` and returning it with [`wp_send_json_error`](https://developer.wordpress.org/reference/functions/wp_send_json_error/)
- The final part of the `handle_request()` method should be the `$this->run()` call.
- The `unlock()` and `next_batch()` lines should be moved to the end of the `run()` method. Reasoning here is that this logic is part of the job running processes. If we were to trigger running a job via a different method (WP_Cron) calling `$this->run()` would still work as expected.
- If you wanted you can use a try/catch block inside the `run()` method to catch any system errors while jobs are running and handle those properly.

Let me know if this makes sense, happy to discuss on a call

Reply:
Made the proposed changes.